plc4j/drivers/opcua/src/main/java/org/apache/plc4x/java/opcua/connection/OpcuaTcpPlcConnection.java

```diff
@@ -0,0 +1,469 @@
+import org.eclipse.milo.opcua.stack.core.types.enumerated.MonitoringMode;
+import org.eclipse.milo.opcua.stack.core.types.enumerated.TimestampsToReturn;
+import org.eclipse.milo.opcua.stack.core.types.structured.EndpointDescription;
+import org.eclipse.milo.opcua.stack.core.types.structured.MonitoredItemCreateRequest;
+import org.eclipse.milo.opcua.stack.core.types.structured.MonitoringParameters;
+import org.eclipse.milo.opcua.stack.core.types.structured.ReadValueId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.math.BigInteger;
+import java.net.InetAddress;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.Unsigned.uint;
+
+public class OpcuaTcpPlcConnection extends BaseOpcuaPlcConnection {
+
+    private static final int OPCUA_DEFAULT_TCP_PORT = 4840;
+
+    private static final Logger logger = LoggerFactory.getLogger(OpcuaTcpPlcConnection.class);
+    private InetAddress address;
+    private int requestTimeout = 5000;
+    private int port;
+    private String params;
+    private OpcUaClient client;
+    private boolean isConnected = false;
+    private  final AtomicLong clientHandles = new AtomicLong(1L);
+
+    private OpcuaTcpPlcConnection(InetAddress address, String params, int requestTimeout) {
+        this( address, OPCUA_DEFAULT_TCP_PORT,  params, requestTimeout);
+        logger.info("Configured OpcuaTcpPlcConnection with: host-name {}", address.getHostAddress());
+    }
+
+    public OpcuaTcpPlcConnection(InetAddress address, int port, String params, int requestTimeout) {
+        this(params);
+        logger.info("Configured OpcuaTcpPlcConnection with: host-name {}", address.getHostAddress());
+        this.address = address;
+        this.port = port;
+        this.params = params;
+    }
+
+    public OpcuaTcpPlcConnection(String params) {
+        super(params);
+    }
+
+    public static OpcuaTcpPlcConnection of(InetAddress address, String params, int requestTimeout) {
+        return new OpcuaTcpPlcConnection(address, params, requestTimeout);
+    }
+
+    public static OpcuaTcpPlcConnection of(InetAddress address, int port, String params, int requestTimeout) {
+        return new OpcuaTcpPlcConnection(address, port, params, requestTimeout);
+    }
+
+
+
+    public InetAddress getRemoteAddress() {
+        return address;
+    }
+
+    @Override
+    public void connect() throws PlcConnectionException {
+        List<EndpointDescription> endpoints =  null;
+
+        try {
+            endpoints = DiscoveryClient.getEndpoints(getEndpointUrl(address, port, params)).get();
```

Comment:
can we have a problem here with infinite timeouts (assume this is a future?)?
Perhaps we should have a default timeout.

Reply:
It will throw a internal Timeout Exception which signals, that the automated discovery was not successfull. But a good point to test it or add a timeout in the get.